### PR TITLE
mysqlproxy: Add auth forwarding for basic auth only

### PR DESCRIPTION
**Pull request by Jeremyyang920**
This commit adds a new auth package that can handle the forwarding of auth packets that will authenticate the client and the remote mysql instance. Note: this only supports basic auth as of now.

### CODE REVIEW

File: server/conn.go [hunk 1]
```diff
@@ -63,16 +62,20 @@ func NewConn(conn net.Conn, user string, password string, h Handler) (*Conn, err
 	}
 	c.closed.Set(false)
 
-	if err := c.handshake(); err != nil {
```
Comment (otan) [inline server/conn.go]: instead of affecting this functionality, can we revert this change?

can we instead use a new method for creating a conn where auth is passthrough? something like `NewAuthedConn(conn net.Conn, h Handler)`

File: proxy/conn.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	}
+}
+
+func (c *Conn) forwardClientPacket() error {
+	clientPacket, err := c.clientConn.ReadPacket()
+	if err != nil {
+		return err
+	}
+
+	err = c.serverConn.WritePacket(append([]byte{0, 0, 0, 0}, clientPacket...))
```
Comment (otan) [inline proxy/conn.go]: can you put a TODO here w.r.t. byte allocation?

File: proxy/conn.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+		return nil, err
+	}
+	err = c.clientConn.WritePacket(append([]byte{0, 0, 0, 0}, mysqlPacket...))
+	if err != nil {
+		return nil, err
+	}
+	return mysqlPacket, nil
+}
+
+func (c *Conn) handleErrorPacket(data []byte) error {
```
Comment (otan) [inline proxy/conn.go:99]: can we re-use the method in `client`, or is it too difficult?

File: proxy/conn.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	}
+	return nil
+}
+
+func (c *Conn) forwardServerPacket() ([]byte, error) {
+	mysqlPacket, err := c.serverConn.ReadPacket()
+	if err != nil {
+		return nil, err
+	}
+	err = c.clientConn.WritePacket(append([]byte{0, 0, 0, 0}, mysqlPacket...))
```
Comment (otan) [inline proxy/conn.go]: nit: `if err := ....; err != nil` here and above

File: proxy/conn.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+			case CACHE_SHA2_FULL_AUTH:
+				err = c.forwardClientPacket()
+				if err != nil {
+					return err
+				}
+
+				mysqlPacket, err = c.forwardServerPacket()
+				if err != nil {
+					return err
+				}
```
Comment (otan) [inline proxy/conn.go:53]: we should have a 
```
default:
   return nil, errors.AssertionFailedf(...)
```

here just in case.

File: proxy/conn.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+		if err != nil {
+			return err
+		}
+
+		if mysqlPacket[0] == OK_HEADER {
+			return nil
+		}
+		if mysqlPacket[0] == ERR_HEADER {
+			return c.handleErrorPacket(mysqlPacket)
+		}
```
Comment (otan) [inline proxy/conn.go]: i think we can delete these as it's already handled by lines 56-63

File: proxy/conn.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+		if mysqlPacket[0] == MORE_DATE_HEADER {
+			switch mysqlPacket[1] {
+			case CACHE_SHA2_FAST_AUTH:
+				mysqlPacket, err = c.forwardServerPacket()
+				if err != nil {
+					return err
+				}
+			case CACHE_SHA2_FULL_AUTH:
+				err = c.forwardClientPacket()
```
Comment (otan) [inline proxy/conn.go]: nit: use `if err = ...` syntax here

**Review by otan — APPROVED** [review]

File: proxy/conn.go (hunk 3 above)
Comment (Jeremyyang920) [reply proxy/conn.go:99]: I wanted to, but there is a capability check in there since they actually unmarshal the packet and stuff so it makes it difficult as we dont have any of the things on that struct.

Comment (otan) [reply proxy/conn.go:99]: something like

```go

func HandleErrorPacket(data []byte) error {
	e := new(MyError)

	var pos = 1

	e.Code = binary.LittleEndian.Uint16(data[pos:])
	pos += 2

	if c.capability&CLIENT_PROTOCOL_41 > 0 {
		//skip '#'
		pos++
		e.State = hack.String(data[pos : pos+5])
		pos += 5
	}

	e.Message = hack.String(data[pos:])

	return e
}

func (*c Conn) handleErrorPacket(data []byte) error {
   return HandleErrorPacket(data)
}
```

don't work?

Comment (otan) [reply proxy/conn.go:99]: oh right, `c.capability`

File: server/conn.go (hunk 1 above)
Comment (Jeremyyang920) [reply server/conn.go]: Yup should be fine to do that. I think only think is we need to expose Handler since its unexported in the current server conn struct.

Comment (Jeremyyang920) [reply server/conn.go]: Nevermind, I just added the method to the server packaged and used it in the new proxy package where we create the authed conn. Think this should do what we want!

**Review by otan — APPROVED** [review]
